File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        visitedA = set()
        
        ptrA = headA
        while ptrA:
            visitedA.add(ptrA)
            ptrA = ptrA.next
        
        ptrB = headB
        while ptrB:
            if ptrB in visitedA:
                return ptrB
            ptrB = ptrB.next
        
        return None

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) return nullptr;

        ListNode *ptr1 = headA, *ptr2 = headB;

        while (ptr1 != ptr2) {
            ptr1 = (ptr1) ? ptr1->next : headB;
            ptr2 = (ptr2) ? ptr2->next : headA;
        }

        return ptr1; // ptr1 == ptr2 at the intersection node or both are nullptr
    }
};
